pkg/cvo/sync_worker.go

```diff
@@ -286,7 +282,13 @@ func (w *SyncWorker) syncPayload(ctx context.Context, work *SyncWork,
 			Update:             desired,
 			LastTransitionTime: time.Now(),
 		})
+
+		// syncPayload executes while locked, but RetrievePayload is a potentially long-running operation
+		// which does not need the lock, so holding it may block other loops (mainly the apply loop) from
+		// execution
+		w.lock.Unlock()
 		info, err := w.retriever.RetrievePayload(ctx, work.Desired)
```

Comment:
This might be a bit risky, because `RetrievePayload` isn't currently re-entrant.  For example, [this cleanup line][1] assumes it's the only currently-running retrieval attempt.  We could add a lock to [the `payloadRetriever` struct][2] or otherwise find a way to deconflict multiple parallel calls to `RetrievePayload`.  Or we could make some kind of argument about how other loops that might be competing for the lock will never be other `syncPayload` rounds that could be trying to call `RetrievePayload`.  I guess we could add a lock to `payloadRetriever` and panic or otherwise complain loudly if we fail to immediately acquire the lock?

We can also slightly reduce the risk of cross-talk here by replacing `work.Desired` (relying on a pointer passed into this function) with `desired` (which we populated from `work.Desired` shorly after entering this function).  If this sounds useful, we could also use `desired.Force` in the klog line a few dozen lines up^^

[1]: https://github.com/openshift/cluster-version-operator/blob/baef28d57e26b3b6a476ef348d7f3274331907be/pkg/cvo/updatepayload.go#L232
[2]: https://github.com/openshift/cluster-version-operator/blob/baef28d57e26b3b6a476ef348d7f3274331907be/pkg/cvo/updatepayload.go#L63-L88

Reply:
Great observation, thanks! I missed this angle. I think locking `payloadRetriever` is a good solution on a first glance but I'll look into it in a bit more detail.

Reply:
I looked into the lock and locked field interactions in CVO, and I believe neither of these two risks applies:

1. CVO does not run `SyncWorker.Update` (and therefore `RetrievePayload`) in parallel
2. `SyncWorker.Update` does not rely on other writers (there is just one, `SyncWorker.Start`) not writing into the locked fields while it executes

Details about my investigation: https://github.com/openshift/cluster-version-operator/pull/1005#issuecomment-1889677070